pkgs/development/interpreters/lua-5/setup-hook.sh

```diff
@@ -0,0 +1,72 @@
+# set -e
+
+nix_print() {
+    if (( "${NIX_DEBUG:-0}" >= $1 )); then
+        echo "$2"
+    fi
+}
+
+nix_debug() {
+    nix_print 3 "$1"
+}
+
+nix_error() {
+    nix_print 1 "$1"
+}
+
+nix_warn() {
+    nix_print 2 "$1"
+}
+addToLuaSearchPathWithCustomDelimiter() {
+    local delimiter="$1"
+    local varName="$2"
+    local pattern="$3"
+    # delete longest match starting from '?'
+    local topDir="${pattern%%\?*}"
+
+    # export only if the folder exists else LUA_PATH grows too big
+    if  [ -d "$topDir" ]; then
+        export "${varName}=${!varName:+${!varName}${delimiter}}${pattern}"
+    fi
+}
+
+addToLuaSearchPath() {
+    addToLuaSearchPathWithCustomDelimiter ";" "$@"
+}
+
+startLuaEnvHook() {
+    addToLuaPath "$1"
+    # echo "LUA_PATH=$LUA_PATH"
+    # echo "LUA_CPATH=$LUA_CPATH"
+    # echo "program_LUA_PATH=$program_LUA_PATH"
+    # echo "program_LUA_CPATH=$program_LUA_CPATH"
+}
+
+# Adds the lib and bin directories to the LUA_PATH and PATH variables,
+# respectively. Recurses on any paths declared in
+# `propagated-native-build-inputs`, while avoiding duplicating paths by
+# flagging the directories it has visited in `luaPathsSeen`.
+addToLuaPath() {
+    local dir="$1"
+    # echo " PATTERNS list $dir  @luapathsearchpaths@"
+
+    if [[ ! -d "$dir" ]]; then
+        echo "$dir not a directory abort"
+        return 0
+    fi
+    cd $dir
+    for pattern in @luapathsearchpaths@;
+    do
+        addToLuaSearchPath LUA_PATH "$PWD/$pattern"
+    done
+
+    # LUA_CPATH
+    for pattern in @luacpathsearchpaths@;
+    do
+        addToLuaSearchPath LUA_CPATH "$PWD/$pattern"
+    done
+    cd -
+}
+
+addEnvHooks "$hostOffset" startLuaEnvHook
```

Comment:
Won't this add paths without checking in `luaPathsSeen`?

Reply:
luaPathsSeen is effective when using the wrapper. The setup-hook is a bit different. With a naive version adding every possible folder for any package, it could end up too long. The version thus checks if the folder exists and only in that case it will add it to LUA_PATH. The "requiredLuaModules" calls `unique` function so it should prevent duplicates from appearing.

Reply:
Are these called for  `requiredLuaModules` or for `buildInputs`, though?

(Also, is there a reasonable way to just use wrappers during normal builds, to reduce code duplication, as well as confusion because of inexact duplication?)

(also, if `luaPathsSeen` is not relevant for this file, why the comment mentions it?)

Reply:
Do you know if any duplication happens in case of multiple buildInputs propagate the same path?

Reply:
I've added a check against duplicates but this hook is run when building user environments so IIRC it is run only on the environment folder.